### ISSUE TITLE
Some tests fail with scikit-learn 0.24

### ISSUE THREAD

**Issue by relf** [SMTorg/smt#274]
Following #272, some tests fail with scikit-learn 0.24. On Windows 7/Python 3.8.3, I get:
<pre>
FAILED smt/tests/test_all.py::Test::test_tanh_GEKPLS_Cobyla - StopIteration
FAILED smt/tests/test_all.py::Test::test_tanh_GEKPLS_TNC - StopIteration
FAILED smt/tests/test_all.py::Test::test_tanh_KPLSK_Cobyla - AssertionError: 0.010782962850413032 not less than or eq...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLSK_TNC - AssertionError: 0.01067608879639801 not less than or equal ...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLS_Cobyla - AssertionError: 0.13695941639924267 not less than or equa...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLS_TNC - AssertionError: 0.1369637566349206 not less than or equal to...
FAILED smt/tests/test_all.py::Test::test_tanh_MGP - AssertionError: 0.13845802478297414 not less than or equal to 0.0201
================================= 7 failed, 41 passed, 1 warning in 113.46s (0:01:53) =================================
</pre>

It seems to be related to changes to PLS which does not work as well as before with tanh function.

**Comment by Paul-Saves** [comment]
Same problems arise with Ubuntu and/or Python 3.9

**Comment by relf** [comment]
It seems that the new implementation of PLS in `scikit-learn 0.24` introduces some kind of regression or at least a behaviour change. It fails when outputs (y) is full of zeros where `scikit-learn 0.23` returns zeroed PLS x rotations (the coefficients we use) and issue a warning. 
```python
import numpy as np
from sklearn.cross_decomposition import PLSRegression
x = np.random.rand(100, 3)
y = np.zeros(100)
pls = PlsRegression.fit(x, y)
print(pls.x_rotations_)
``` 
with 0.24.1
```bash
Traceback (most recent call last):
  File "test_bug.py", line 7, in <module>
    pls = PLSRegression().fit(x, y)
  File "D:\Miniconda3\lib\site-packages\sklearn\cross_decomposition\_pls.py", line 243, in fit
    _get_first_singular_vectors_power_method(
  File "D:\Miniconda3\lib\site-packages\sklearn\cross_decomposition\_pls.py", line 37, in _get_first_singular_ve
    y_score = next(col for col in Y.T if np.any(np.abs(col) > eps))
StopIteration
```
with scikit-learn 0.23.2
```bash
D:\Miniconda3\lib\site-packages\sklearn\cross_decomposition\_pls.py:323: UserWarning: Y residual constant at iteration 0
warnings.warn('Y residual constant at iteration %s' % k)
[[0. 0.]
 [0. 0.]
 [0. 0.]]
```
See also https://github.com/scikit-learn/scikit-learn/issues/19831

I am going to fix our code to restore the sklearn 0.23 behaviour (ie zeroed x_rotations when StopIteration is raised by PLS fit).

**Comment by relf** [comment]
PLS regression was refactored in `scikit-learn 0.24`, one change is the raising of StopIteration exception but there are also cases where `x_rotations_` returned values are quite different. 

In the case of our tanh KPLS Cobyla test, with `numpy 1.20.2`, `scipy 1.6.2` I get the following values for `pls.x_rotations_`  : 
* on Linux
  * with 0.23.2 <code>[[0.36136663], [0.3759673 ], [0.36866696]]</code>
  * with 0.24.1 <code>[[0.        ], [0.03865775], [0.01932888]]</code>
* On windows 7, (tanh tests are ok!)
  * with 0.23.2 <code>[[0.38882051], [0.3742435 ], [0.34258016]]</code>
  * with 0.24.1<code>[[ 0.07066039], [ 0.0339257 ], [-0.04586723]]</code>

When I look at the training data (generated with FullFactorial DOE), it appears there are "many" zeros hence the tensorproduct of tanh produces also more zeros. 

So far we've used `FullFactorial` to get reproducible test results, but for some time now we have a random seed for LHS, I've just tried with it and the errors have just vanished. 

It remains that `scikit-learn 0.24.1` introduces changes in PLS methods that impact KPLS family results, specially when many zeros (or even constants?) are present in the training data.
